protocols/relay/src/v2/client/handler.rs

```diff
@@ -0,0 +1,610 @@
+                                self.pending_error = Some(ProtocolsHandlerUpgrErr::Upgrade(
+                                    upgrade::UpgradeError::Apply(EitherError::B(error)),
+                                ));
+                            }
+                            outbound_hop::UpgradeError::UnexpectedStatus(status) => {
+                                match status {
+                                    Status::Ok => {
+                                        unreachable!(
+                                            "Status success was explicitly expected earlier."
+                                        )
+                                    }
+                                    // With either status below there is either no reason to stay
+                                    // connected or it is a protocol violation.
+                                    // Thus terminate the connection.
+                                    Status::ReservationRefused
+                                    | Status::UnexpectedMessage
+                                    | Status::MalformedMessage => {
+                                        self.pending_error =
+                                            Some(ProtocolsHandlerUpgrErr::Upgrade(
+                                                upgrade::UpgradeError::Apply(EitherError::B(error)),
+                                            ));
+                                    }
+                                    // While useless for reaching this particular destination, the
+                                    // connection to the relay might still proof helpful for other
+                                    // destinations. Thus do not terminate the connection.
+                                    Status::ResourceLimitExceeded
+                                    | Status::ConnectionFailed
+                                    | Status::NoReservation
+                                    | Status::PermissionDenied => {}
+                                }
+                            }
+                        }
+                    }
+                };
+
+                let _ = send_back.send(Err(()));
+
+                self.queued_events.push_back(ProtocolsHandlerEvent::Custom(
+                    Event::OutboundCircuitReqFailed {},
+                ));
+            }
+        }
+    }
+
+    fn connection_keep_alive(&self) -> KeepAlive {
+        self.keep_alive
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ProtocolsHandlerEvent<
+            Self::OutboundProtocol,
+            Self::OutboundOpenInfo,
+            Self::OutEvent,
+            Self::Error,
+        >,
+    > {
+        // Check for a pending (fatal) error.
+        if let Some(err) = self.pending_error.take() {
+            // The handler will not be polled again by the `Swarm`.
+            return Poll::Ready(ProtocolsHandlerEvent::Close(err));
```

Comment:
Yeah that makes sense. I wrote a bit more thoughts here: https://github.com/libp2p/rust-libp2p/pull/2059#discussion_r764317943.

I think this makes sense. This handler is in an unexpected state so we should just stop this handler. My only questions would then be:

1. We aren't relying on the errors propagating to the transport then? (since we may/will drop some)
2. The swarm drops this object, correct?
3. The NetworkBehavior still gets a `inject_disconnected/inject_connection_closed`, correct?

Reply:
2 & 3 happen here I believe: https://github.com/libp2p/rust-libp2p/blob/master/core/src/connection/pool/task.rs#L244

Reply:
Resolving here, see https://github.com/libp2p/rust-libp2p/pull/2059#discussion_r765832463. Let me know in case I am missing something.